cmd/db_repair/repair.go

```diff
@@ -232,25 +212,174 @@ func (pc *PostgresClient) RepairEvents(cfClient cf.Client) (err error) {
 		if err != nil {
 			return err
 		}
-		fmt.Println("adding missing app CREATED event", ev)
 		_, err = tx.Exec(`
 			insert into service_usage_events (
-				id
+				id,
 				guid,
 				created_at,
 				raw_message
 			) values (
-				0,
-				$1::text,
-				$2::timestamp,
-				$3::jsonb
+				$1,
+				$2::text,
+				$3::timestamp,
+				$4::jsonb
 			)
-		`, serviceInstance.Guid, firstEventTimestamp, string(evJSON))
+		`, fakeID, serviceInstance.Guid, firstEventTimestamp, string(evJSON))
 		if err != nil {
 			return err
 		}
+		fakeEventsInserted++
 		knownGuids[serviceInstance.Guid] = true
 	}
+	fmt.Println("Fake service CREATED events inserted:", fakeEventsInserted)
+	return nil
+}
+
+func deleteFakeEvents(tx *sql.Tx) (err error) {
+	// remove any events with fake ID
+	_, err = tx.Exec(`delete from app_usage_events where id = $1`, fakeID)
+	if err != nil {
+		return err
+	}
+	_, err = tx.Exec(`delete from service_usage_events where id = $1`, fakeID)
+	if err != nil {
+		return err
+	}
+	return nil
+}
 
+func getCollectionEpoch(tx *sql.Tx) (*time.Time, error) {
+	var firstEventTimestamp *time.Time
+	err := tx.QueryRow(`
+		select least((
+			select min(created_at::timestamptz) from app_usage_events where created_at is not null
+		),(
+			select min(created_at::timestamptz) from service_usage_events where created_at is not null
+		));
+	`).Scan(&firstEventTimestamp)
+	if err != nil {
+		return nil, err
+	}
+	if firstEventTimestamp == nil {
+		return nil, errors.New("Database appears to be empty and thus cannot be repaired.")
+	}
+	return firstEventTimestamp, nil
+}
+
+func createEventsForAppsWhereFirstRecordedEventIsStopped(tx *sql.Tx, firstEventTimestamp *time.Time) (err error) {
+	result, err := tx.Exec(`
+		WITH events AS (
```

Comment:
Not important: this seems quite expensive to me as for one application there will be multiple lines. Is there a way to use group by here for the app_guids and simply select the first raw_message (ordered by id)

Reply:
Could you elaborate on 'simply select the first raw_message' please? What would be doing the selecting? It would be good if there was a less expensive way of doing this.

Reply:
I don't know Postgres as well. After a quick google it seems we could use the row_number() function on the window and include a condition in where to include only the lines with row_number = 1

Reply:
https://stackoverflow.com/a/33391656/873178

Reply:
You could get the same effect by just combining the `events` and `stop_events` CTEs into a single query.... but we had split them out this way for readability over performance since this is a very infrequently run task :)